diesel/src/query_builder/upsert/on_conflict_target_decorations.rs

```diff
@@ -0,0 +1,68 @@
+use crate::backend::{Backend, SupportsOnConflictClause};
+use crate::expression::Expression;
+use crate::query_builder::upsert::on_conflict_target::{ConflictTarget, NoConflictTarget};
+use crate::query_builder::where_clause::{NoWhereClause, WhereAnd, WhereClause};
+use crate::query_builder::{AstPass, QueryFragment, QueryResult};
+use crate::sql_types::Bool;
+
+pub trait UndecoratedConflictTarget {}
+
+impl UndecoratedConflictTarget for NoConflictTarget {}
+impl<T> UndecoratedConflictTarget for ConflictTarget<T> {}
+
+/// Interface to add information to conflict targets.
+/// Designed to be open for further additions to conflict targets like constraints
+pub trait DecoratableTarget<T, U, P> {
+    /// Output type of filter_target operation
+    type FilterOutput;
+    /// equivalent to filter of FilterDsl but aimed at conflict targets
+    fn filter_target(self, predicate: P) -> Self::FilterOutput;
+}
+
+#[derive(Debug)]
+pub struct DecoratedConflictTarget<T, U> {
+    target: T,
+    where_clause: U,
+}
+
+impl<T, P> DecoratableTarget<T, WhereClause<P>, P> for T
+where
+    P: Expression<SqlType = Bool>,
+    T: UndecoratedConflictTarget,
+{
+    type FilterOutput = DecoratedConflictTarget<T, WhereClause<P>>;
+    fn filter_target(self, predicate: P) -> Self::FilterOutput {
+        DecoratedConflictTarget {
+            target: self,
+            where_clause: NoWhereClause.and(predicate),
+        }
+    }
+}
+
+impl<T, U, P> DecoratableTarget<T, <U as WhereAnd<P>>::Output, P> for DecoratedConflictTarget<T, U>
+where
+    P: Expression<SqlType = Bool>,
+    U: WhereAnd<P>,
+{
+    type FilterOutput = DecoratedConflictTarget<T, <U as WhereAnd<P>>::Output>;
+    fn filter_target(self, predicate: P) -> Self::FilterOutput {
+        DecoratedConflictTarget {
+            target: self.target,
+            where_clause: self.where_clause.and(predicate),
+        }
+    }
+}
+
+impl<DB, T, U> QueryFragment<DB> for DecoratedConflictTarget<T, U>
+where
+    T: QueryFragment<DB>,
+    U: QueryFragment<DB>,
+    DB: Backend + SupportsOnConflictClause,
+{
+    fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+        self.target.walk_ast(out.reborrow())?;
+        // out.push_sql(" ");
```

Comment:
Is this a debug statement left in that is no longer effectual?

Reply:
oh yeah, that's just a remnant. I'll remove it.